Include software version in URLs for style and javascript files.

bmbroom opened the pull request:
This is a start on a possible way to address the client caching problem on version changes.

The intention is to override client and proxy caching when **and only when**
the software version changes.  Basically, the version number is included
in the URLs for style and javascript files. It requires the web server to ignore those
version numbers.  Example rewrite rules for NGINX are included.

This commit breaks compilation of source files into widget or standalone
artifacts.  If this seems a reasonable way forward, we can address those
issues in future commits.

marohrdanz commented:
Just adding another idea, based on our meeting discussion.

This idea is less precise/complete than this pull request, but is also less complex. Maybe we could use somthing like the following in a .htaccess file? (assuming apache httpd with [mod_expires](https://httpd.apache.org/docs/2.4/mod/mod_expires.html) enabled rather than nginx)

```apache
<IfModule mod_expires.c>
    ExpiresActive On
    ExpiresByType image/png "access plus 1 day"
    ExpiresByType text/javascript "access plus 1 day" 
    ExpiresByType application/javascript "access plus 1 day" 
    ExpiresByType text/css "access plus 1 day" 
    ExpiresDefault "access plus 1 day" 
</IfModule>
```

jmelott commented:
Looking at the code changes, this would require the version number to be changed in two places each time a new version is released and would only update the code in the viewer.  
Once in NGCHM/WebContent/chm.html and once in NGCHM/WebContent/javascript/CompatibilityManager.js

We'd probably need to something similar in ngChmApp.html and/or Select_Matrix.html in the GUI Builder
If they call their own css or JavaScript files then Cluster_Matrix.html, Edit_Covariates.html, Format_Display.html, Transform_Matrix.html, and View_HeatMap.html may also need similar alterations.

There might also need to be similar alterations on multiple pages to the the GenePattern Builder, Galaxy Basic and Advanced Builders, and Advanced (DUI) Builder.  It would be better if we could come up with a way to do this where the version number doesn't need to be changed in so many individual places as they will eventually fall out of sync unless there is an automated manner of changing all of them.

For those applications that are using a webserver to serve the application, like those listed above, could we perhaps do a hybrid of this and Mary's suggestion. Make a new file type on the web server that just holds a version number, say txt/version_id.  Add an entry on the webserver to say to cache that file type infrequently or never.  Modify the script routines added to load that version id from the website every time, then use that value to get and cache the latest version of all the other files.  We'd only need to change the version number in one file per webserver server and one could also use that to find out what version of the software is supposedly running on that server without having to run the program and look at an about dialog or other parts of the code manually.

In both cases this won't solve the issue with the standalone viewer and there might need to be some tweaking done where people using the embedded viewer on their websites might need to include such changes on their server or they would still get whatever version was in the user's browser cache or whatever is on their system.

Single-page HTML maps I guess would never be updated if all the code is incorporated into the web page and never goes to the server.

mstucky commented:
These are some great inputs and ideas to be considered.  At the Thursday we agreed to address this issue as a PT item for a future release.  It would then be a part of 2.20.x and we would have the time to truly examine the issue in detail.  This would include the impact on all our subsidiary products (Viewer, Builder, Embedded, Standalone, Galaxy).

bmbroom commented:
This isn't (or shouldn't be) true:

Once in NGCHM/WebContent/chm.html and once in NGCHM/WebContent/javascript/CompatibilityManager.js

I removed the version number definition from CompatibilityManager and added it to chm.html.

mstucky commented:
One thing I would note from looking at this change.  It will most definitely break the process whereby we build the widgetized ngchm js file.  That process loads in each JS file by working off of chm.html.  This does not mean that the process cannot be changed but it will have to for this to work.
